### ISSUE TITLE
Exploit: Using a macro can get you 40-100 free wishes

### ISSUE THREAD

**Issue by s1h4d0w** [Rsl1122/Genie#22]
I've had to remove the plugin from my server because some players found out that if you use a macro to spam a wish to chat you can go far beyond the wishes your genie has. Apparently the plugin has no proper check to see if a lamp runs out if you spam wishes quick enough.

Players just spam the chat with 10-100x "I want a \<item\>" within a second and the plugin actually grants the wish that many times, even though their lamp only has 3 wishes.

Proof, from the "lamps.yml":
![image](https://user-images.githubusercontent.com/5654654/113415374-5a6c0f80-93bf-11eb-95d6-d836840fb9a3.png)
![image](https://user-images.githubusercontent.com/5654654/113420384-82607080-93c9-11eb-9059-fb6ee79e811c.png)

**Comment by AuroraLS3** [comment]
Interesting. Seems like they managed to bypass [the if block](https://github.com/AuroraLS3/Genie/blob/master/Genie/src/main/java/com/djrapitops/genie/listeners/ChatListener.java#L54) since each chat event runs on its own thread and is independent of each other..

**Comment by s1h4d0w** [comment]
Any idea if there's an easy fix for that? Maybe queuing wishes and running them one by one instead of immediately after the chat message event?

**Comment by AuroraLS3** [comment]
Yeah it's essentially a concurrency issue so either moving them to a single thread or adding a resource control (like a semaphore/locking) should sort it out.
